components/brave_wallet/browser/ethereum_permission_utils.cc

```diff
@@ -0,0 +1,158 @@
+// end of the host piece of the origin, then return it as the new origin.
+bool AddAccountToHost(const GURL& old_origin,
+                      const std::string& account,
+                      GURL* new_origin) {
+  if (!old_origin.is_valid() || account.empty() || !new_origin)
+    return false;
+
+  GURL::Replacements replacements;
+  std::string new_host = base::StrCat({old_origin.host_piece(), account});
+  replacements.SetHostStr(new_host);
+
+  *new_origin = old_origin.ReplaceComponents(replacements);
+  return new_origin->is_valid();
+}
+
+// Given the overwritten origin, such as https://test.com{addr=123&addr=456},
+// extract all addresses and save into address_queue.
+void ExtractAddresses(const GURL& origin,
+                      std::queue<std::string>* address_queue) {
+  static const base::NoDestructor<re2::RE2> kAddrRegex(kAddrPattern);
+  DCHECK(origin.is_valid() && address_queue);
+  DCHECK_EQ(origin, origin.GetOrigin());
+
+  re2::StringPiece input(origin.spec());
+  std::string match;
+  while (re2::RE2::FindAndConsume(&input, *kAddrRegex, &match)) {
+    address_queue->push(match);
+  }
+}
+
+// Parse requesting origin in either sub-request format (one address) or
+// non-sub-request format (all addresses).
+bool ParseRequestingOriginInternal(const GURL& origin,
+                                   bool sub_req_format,
+                                   std::string* requesting_origin,
+                                   std::string* account,
+                                   std::queue<std::string>* address_queue) {
+  if (!origin.is_valid() || origin != origin.GetOrigin())
+    return false;
+
+  std::string scheme_host_group;
+  std::string address_group;
+  std::string port_group;
+
+  // Validate input format.
+  std::string pattern = sub_req_format
+                            ? "(.*)(0x[[:xdigit:]]{40})(:[0-9]+)*/"
+                            : "(.*)%7Baddr%3D0x[[:xdigit:]]{40}(%"
+                              "26addr%3D0x[[:xdigit:]]{40})*%7D(:[0-9]+)*/";
+  RE2 full_pattern(pattern);
+  if (!re2::RE2::FullMatch(origin.spec(), full_pattern, &scheme_host_group,
+                           &address_group, &port_group)) {
+    return false;
+  }
+
+  if (requesting_origin) {
+    *requesting_origin = base::StrCat({scheme_host_group, port_group});
+  }
+
+  if (sub_req_format && account) {
+    *account = address_group;
+  }
+
+  if (!sub_req_format && address_queue) {
+    ExtractAddresses(origin, address_queue);
+  }
+
+  return true;
+}
+
+}  // namespace
+
+namespace brave_wallet {
+
+bool GetConcatOriginFromWalletAddresses(
+    const GURL& old_origin,
+    const std::vector<std::string>& addresses,
+    GURL* new_origin) {
+  if (!old_origin.is_valid() || addresses.empty()) {
+    return false;
+  }
+
+  std::string addresses_suffix = "{";
+  for (const auto& address : addresses) {
+    base::StrAppend(&addresses_suffix, {"addr=", address, "&"});
```

Comment:
nit: I would prefer an iterator that checks `!= it.end() - 1` for appending `"&"` over popping it back off at the end. I know you have a check for `addresses.empty` above, but it still seems better to ensure this works even if it is empty

Reply:
This is fixed in https://github.com/brave/brave-core/pull/9443/commits/d49d9bfcd494049768ef0a4c9945a46430268567